File: apps/web/src/app/features/upl/formats/upl-format-model.ts
import { UplColumn, UplFormatDraftRequest, UplSheet } from '../upl-api';
import { UplFieldError } from './upl-format-errors';

/** Шаги анкеты формата. */
export type UplFormatStep = 'file' | 'sheets' | 'publish';

/** Поля верхнего уровня, которые правятся на шаге «Файл». */
const FILE_STEP_FIELDS = new Set(['fileKind', 'encoding', 'delimiter', 'matchColumnsBy']);

/** Пустая строка в необязательном поле означает «не заполнено», а не пустое значение. */
export function trimToNull(value: string | null | undefined): string | null {
  const trimmed = (value ?? '').trim();
  return trimmed.length === 0 ? null : trimmed;
}

export function isFilled(value: string | number | null | undefined): boolean {
  return value !== null && value !== undefined && value !== '';
}

export function emptyModel(): UplFormatDraftRequest {
  return { lockVersion: 0, fileKind: 'xlsx', encoding: null, delimiter: null, matchColumnsBy: 'header', sheets: [] };
}

export function emptySheet(): UplSheet {
  return { id: null, ordinal: null, sheetName: null, headerRow: 1, totalRowMarker: null, columns: [] };
}

export function emptyColumn(): UplColumn {
  return {
    id: null,
    ordinal: null,
    filePosition: null,
    nameInFile: '',
    targetField: '',
    dataType: 'text',
    required: false,
    sourceUnit: null,
    baseUnit: null,
    keyMask: null,
    keyPadLength: null,
    keyPadMax: null,
    refBookCode: null
  };
}

export function isNumericColumn(column: UplColumn): boolean {
  return column.dataType === 'integer' || column.dataType === 'number';
}

/** Очищает поля, которых у текущего типа колонки нет; true — если что-то было заполнено. */
export function clearFieldsForType(column: UplColumn): boolean {
  let cleared = false;
  if (!isNumericColumn(column) && (isFilled(column.sourceUnit) || isFilled(column.baseUnit))) {
    column.sourceUnit = null;
    column.baseUnit = null;
    cleared = true;
  }
  if (column.dataType !== 'object_key'
    && (isFilled(column.keyMask) || isFilled(column.keyPadLength) || isFilled(column.keyPadMax))) {
    column.keyMask = null;
    column.keyPadLength = null;
    column.keyPadMax = null;
    cleared = true;
  }
  if (column.dataType !== 'ref_code' && isFilled(column.refBookCode)) {
    column.refBookCode = null;
    cleared = true;
  }
  return cleared;
}

/** На каком шаге правится поле с ошибкой: адрес листа — «Листы и колонки», поля файла — «Файл». */
export function uplErrorStep(error: UplFieldError): UplFormatStep {
  return error.sheet === null && FILE_STEP_FIELDS.has(error.field) ? 'file' : 'sheets';
}

File: apps/server/src/main/java/com/greenwhite/dwh/instance/upl/format/UplFormatModel.java
package com.greenwhite.dwh.instance.upl.format;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.Locale;

/** Модель анкеты файла: источник, версия формата, листы и колонки (таблицы V111). */
public final class UplFormatModel {

    private UplFormatModel() {
    }

    private static String toDb(Enum<?> value) {
        return value.name().toLowerCase(Locale.ROOT);
    }

    private static <E extends Enum<E>> E parse(Class<E> type, String value) {
        if (value == null) {
            throw new IllegalArgumentException("Пустое значение для " + type.getSimpleName());
        }
        return Enum.valueOf(type, value.toUpperCase(Locale.ROOT));
    }

    public enum Periodicity {
        MONTH, QUARTER, YEAR, ADHOC;

        public String db() {
            return toDb(this);
        }

        public static Periodicity fromDb(String v) {
            return parse(Periodicity.class, v);
        }
    }

    public enum SourceType {
        FILE;

        public String db() {
            return toDb(this);
        }

        public static SourceType fromDb(String v) {
            return parse(SourceType.class, v);
        }
    }

    public enum Strictness {
        ERROR, WARNING;

        public String db() {
            return toDb(this);
        }

        public static Strictness fromDb(String v) {
            return parse(Strictness.class, v);
        }
    }

    public enum FileKind {
        XLSX, CSV;

        public String db() {
            return toDb(this);
        }

        public static FileKind fromDb(String v) {
            return parse(FileKind.class, v);
        }
    }

    public enum MatchBy {
        HEADER, POSITION;

        public String db() {
            return toDb(this);
        }

        public static MatchBy fromDb(String v) {
            return parse(MatchBy.class, v);
        }
    }

    public enum DataType {
        TEXT, INTEGER, NUMBER, DATE, OBJECT_KEY, REF_CODE;

        public String db() {
            return toDb(this);
        }

        public static DataType fromDb(String v) {
            return parse(DataType.class, v);
        }
    }

    public record SourceData(String code, String name, String ownerOrg, String ownerContact,
                             Periodicity periodicity, int slaDays, SourceType sourceType, Strictness strictness) {
    }

    public record Source(long id, String code, String name, String ownerOrg, String ownerContact,
                         Periodicity periodicity, int slaDays, SourceType sourceType, Strictness strictness,
                         int lockVersion, Instant createdAt, String createdBy, Instant modifiedAt, String modifiedBy) {
    }

    public record SourceSummary(long id, String code, String name, Periodicity periodicity,
                                Integer lastPublishedVersion, boolean hasDraft) {
    }

    /**
     * @param headerSynonyms other headers the file may carry for this column (matched like {@code nameInFile})
     */
    public record Column(Long id, int ordinal, Integer filePosition, String nameInFile, String targetField,
                         DataType dataType, boolean required, String sourceUnit, String baseUnit,
                         String keyMask, Integer keyPadLength, Integer keyPadMax, String refBookCode,
                         List<String> headerSynonyms) {

        public Column {
            headerSynonyms = headerSynonyms == null ? List.of() : List.copyOf(headerSynonyms);
        }

        /** A column without synonyms. */
        public Column(Long id, int ordinal, Integer filePosition, String nameInFile, String targetField,
                      DataType dataType, boolean required, String sourceUnit, String baseUnit,
                      String keyMask, Integer keyPadLength, Integer keyPadMax, String refBookCode) {
            this(id, ordinal, filePosition, nameInFile, targetField, dataType, required, sourceUnit, baseUnit,
                    keyMask, keyPadLength, keyPadMax, refBookCode, List.of());
        }

        /** Every header this column accepts: its name first, then the synonyms. */
        public List<String> acceptedHeaders() {
            List<String> all = new java.util.ArrayList<>();
            all.add(nameInFile);
            all.addAll(headerSynonyms);
            return all;
        }
    }

    public record Sheet(Long id, int ordinal, String sheetName, int headerRow, String totalRowMarker,
                        List<Column> columns) {
    }

    public record FormatVersion(long sourceId, int version, LocalDate validFrom, LocalDate validTo, String status,
                                Instant publishedAt, String publishedBy, int lockVersion, FileKind fileKind,
                                String encoding, String delimiter, MatchBy matchColumnsBy, List<Sheet> sheets) {

        public FormatVersion withSheets(List<Sheet> s) {
            return new FormatVersion(sourceId, version, validFrom, validTo, status, publishedAt, publishedBy,
                    lockVersion, fileKind, encoding, delimiter, matchColumnsBy, s);
        }
    }
}
